### ISSUE TITLE
Is the generated DLL complete, or is more needed to call code (from Metatrader 4)?

### ISSUE THREAD

**Issue by DennisGundersen** [AaronRobinsonMSFT/DNNE#165]
I've made a test project, but my client (MetaTrader 4) just collapses on the first method call (DLL does load successfully), so there's no feedback on what's happening. I was wondering if maybe the .h files has to be included somehow, but importing those just adds more calls for additional inclusions of stddef.h and stdint.h. Maybe it's the JSON files? I have virtually no skills in C++, but it feels to me that this is some sort of C++ connecting issue. Anything pop out? Could be something really simple that "everyone knows, but me".

The csproj is:
`<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup Label="Project">
    <TargetFramework>net7.0-windows</TargetFramework>
    <Nullable>enable</Nullable>
    <Platforms>x86</Platforms>
    <PlatformTarget>x86</PlatformTarget>
    <RuntimeIdentifier>win-x86</RuntimeIdentifier>
    <EnableDynamicLoading>true</EnableDynamicLoading>
    <OutputType>library</OutputType>
  </PropertyGroup>
  <PropertyGroup Label="DNNE">
    <DnneSelfContained_Experimental>true</DnneSelfContained_Experimental>
    <DnneAddGeneratedBinaryToProject>true</DnneAddGeneratedBinaryToProject>
    <AutoGenerateBindingRedirects>true</AutoGenerateBindingRedirects>
    <InvariantGlobalization>true</InvariantGlobalization>
  </PropertyGroup>
  <PropertyGroup Label="Publish">
    <GenerateRuntimeConfigurationFiles>true</GenerateRuntimeConfigurationFiles>
    <RollForward>LatestMinor</RollForward>
    <PublishSingleFile>false</PublishSingleFile>
    <SelfContained>false</SelfContained>
    <PublishTrimmed>false</PublishTrimmed>
    <DebugType>embedded</DebugType>
    <IncludeNativeLibrariesForSelfExtract>false</IncludeNativeLibrariesForSelfExtract>
    <ErrorOnDuplicatePublishOutputFiles>false</ErrorOnDuplicatePublishOutputFiles>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="DNNE" Version="2.0.1" />
  </ItemGroup>
</Project>`

and the class is simply:

`using System.Runtime.InteropServices;
namespace BridgeUsingDNNEForMT4
{
    public static class DNNEBridge
    {
        [UnmanagedCallersOnly(EntryPoint = "GimmeAnInt")]
        public static int GimmeAnInt(int a)
        {
            return a + 1;
        }
    }
}
`

MT4 uses a language called MQL, which is a subset of C++ (apparently), to run external code. This is my "expert adviser" (name for  compiled programs in MT4):

`//+------------------------------------------------------------------+
//|                                                     DNNETest.mq4 |
//|                                  Copyright 2023, MetaQuotes Ltd. |
//|                                             https://www.mql5.com |
//+------------------------------------------------------------------+
#property copyright "Copyright 2023, MetaQuotes Ltd."
#property link      "https://www.mql5.com"
#property version   "1.00"
#property strict

//#include <BridgeUsingDNNEForMT4NE.h>

#import "BridgeUsingDNNEForMT4NE.dll"
   int GimmeAnInt(int a);
#import

//+------------------------------------------------------------------+
//| Expert initialization function                                   |
//+------------------------------------------------------------------+
int OnInit()
  {
   int value =  2;
   int result = GimmeAnInt(value);
   Print("GimmeAnInt(" + IntegerToString(value) + ") returned: " + IntegerToString(result));
   return(INIT_SUCCEEDED);
  }
`

TIA!
Dennis

**Comment by AaronRobinsonMSFT** [comment]
**TL;DR** Update project to `<DnneSelfContained_Experimental>false</DnneSelfContained_Experimental>` or ideally remove that entirely. It is experimental and not worth fighting unless actually needed.

>  just collapses on the first method call (DLL does load successfully), so there's no feedback on what's happening.

@DennisGundersen Yep. DNNE calls C's `abort` if something it can't recover from happens. The FAQ has guidance on how to customize this.

Just looking at the project, I see two conflicting ideas.

```xml
    <DnneSelfContained_Experimental>true</DnneSelfContained_Experimental>
...
    <PublishSingleFile>false</PublishSingleFile>
    <SelfContained>false</SelfContained>
```

Asking DNNE to try self-contained, but then telling the .NET SDK to _not_ use self-contained is going to be very confusing. My first pass would avoid self-contained entirely and get the typical framework dependent working. After that you can experiment with more advanced features.

**Comment by DennisGundersen** [comment]
> Asking DNNE to try self-contained, but then telling the .NET SDK to _not_ use self-contained is going to be very confusing. My first pass would avoid self-contained entirely and get the typical framework dependent working. After that you can experiment with more advanced features.

Okay, I think I get that... The original props were:
`    <DnneSelfContained_Experimental>true</DnneSelfContained_Experimental>
    <PublishSingleFile>false</PublishSingleFile>
    <SelfContained>true</SelfContained>
    <PublishTrimmed>true</PublishTrimmed>`

That gives me projectNE.DLL + plus framework files, but it crashes

Changing it to:
`    <DnneSelfContained_Experimental>false</DnneSelfContained_Experimental>
    <PublishSingleFile>false</PublishSingleFile>
    <SelfContained>false</SelfContained>
    <PublishTrimmed>true</PublishTrimmed>`

fails on publish due to "Optimizing assemblies for size is not supported for the selected publish configuration. Please ensure that you are publishing a self-contained app."

I therefore end up with:
`  <PropertyGroup Label="DNNE">
    <DnneSelfContained_Experimental>false</DnneSelfContained_Experimental>
    <DnneAddGeneratedBinaryToProject>true</DnneAddGeneratedBinaryToProject>
    <AutoGenerateBindingRedirects>true</AutoGenerateBindingRedirects>
    <InvariantGlobalization>true</InvariantGlobalization>
  </PropertyGroup>
  <PropertyGroup Label="Publish">
    <GenerateRuntimeConfigurationFiles>true</GenerateRuntimeConfigurationFiles>
    <RollForward>LatestMinor</RollForward>
    <PublishSingleFile>false</PublishSingleFile>
    <SelfContained>false</SelfContained>
    <PublishTrimmed>false</PublishTrimmed>
    <DebugType>embedded</DebugType>
    <IncludeNativeLibrariesForSelfExtract>false</IncludeNativeLibrariesForSelfExtract>
    <ErrorOnDuplicatePublishOutputFiles>false</ErrorOnDuplicatePublishOutputFiles>
  </PropertyGroup>`

which gives me my projectNE.dll, no framework files, but also crashes.

I've also tried to update the export decorations to:
` [UnmanagedCallersOnly(CallConvs = new[] { typeof(System.Runtime.CompilerServices.CallConvCdecl) })]`
in case MT4 doesn't know what an int is, but it doesn't seem to make a difference.

Am I still doing DNNE wrong, or is it MT4?

Re
Dennis

**Comment by AaronRobinsonMSFT** [comment]
@DennisGundersen Could be a few things. I assume you are building the DNNE assembly in Debug mode?

Is it possible to debug the MQL? If so, I would "step into" on the line below and you can see where the code is failing. Keep stepping in until you get to [`prepare_runtime()`](https://github.com/AaronRobinsonMSFT/DNNE/blob/cd55d290c56c022a7a73b48f0ecf232546bee6e3/src/platform/platform.c#L488-L511) where you can likely narrow down the underlying issue.

```cpp
//+------------------------------------------------------------------+
//| Expert initialization function                                   |
//+------------------------------------------------------------------+
int OnInit()
  {
   int value =  2;
   int result = GimmeAnInt(value); <----- Step in here.
   Print("GimmeAnInt(" + IntegerToString(value) + ") returned: " + IntegerToString(result));
   return(INIT_SUCCEEDED);
  }
```

**Comment by DennisGundersen** [comment]
> Is it possible to debug the MQL? If so, I would "step into" on the line below and you can see where the code is failing. Keep stepping in until you get to [`prepare_runtime()`]

Unfortunately, not for external code, only inside the MQL itself. And it only has one error code.
Programming is fun, isn't it?

Re
Dennis

**Comment by AaronRobinsonMSFT** [comment]
> Unfortunately, not for external code, only inside the MQL itself. And it only has one error code.
> Programming is fun, isn't it?

Sad. Let's try something else then.

1) Use [`ProcMon`](https://learn.microsoft.com/sysinternals/downloads/procmon) to see what files are being examined. This plugs into the file system so you should be able to see all file system access for the process in question.

2) You can override the `abort()` call. This is described in the [FAQs](https://github.com/AaronRobinsonMSFT/DNNE#faqs). In order overriden version of `abort()`, you can trigger a breakpoint using [`DebugBreak()`](https://learn.microsoft.com/windows/win32/api/debugapi/nf-debugapi-debugbreak) or even create a simple loop that will spin until you can attach a debugger like Visual Studio or [WinDBG](https://learn.microsoft.com/windows-hardware/drivers/debugger/debugger-download-tools). From there, you should be able to see the stack and perhaps error code looking up the stack.

**Comment by ThaDaVos** [comment]
This is my project:
https://github.com/thadavos/experimental-calculator

**Comment by DennisGundersen** [comment]
> @DennisGundersen - if I remember correctly, you're using my template, right?
> 
Yes, I'm using the @ThaDaVos template, and I have the same publishing setup. So I guess that means it must be a Metatrader issue then? I've asked an actual programmer (who knows MT4) to help me, so hopefully there will be some clarifications, or at least better questions soon.

Re
Dennis

**Comment by ThaDaVos** [comment]
Something cool I just figured out - you can remove the:
```XML
<Platforms>x86</Platforms>
<PlatformTarget>x86</PlatformTarget>
<RuntimeIdentifier>win-x86</RuntimeIdentifier>
```

And use only
```XML
<DnneRuntimeIdentifier>win-x86</DnneRuntimeIdentifier>
```

This will make your Library DLL be AnyCPU but the DNNE native library be targeted to `win-x86` - it works for me in Clarion - also, it seems to be easier to deal with

**Comment by DennisGundersen** [comment]
> ```
> <DnneRuntimeIdentifier>win-x86</DnneRuntimeIdentifier>
> ```
> 
> This will make your Library DLL be AnyCPU but the DNNE native library be targeted to `win-x86` - it works for me in Clarion - also, it seems to be easier to deal with

MT4 doesn't work with AnyCPU, only x86, but thanks for the suggestion.

Re
Dennis

**Comment by ThaDaVos** [comment]
MT4 will be using the x86 NE dll - the actual project DLL will be AnyCPU.

About the dependencies - this, as far as I know, has always been this way - as the default DotNet install does not contain Nuget Package dll's and stuff.

One thing you could do - which is also what I am doing - is use `Fody` and `Fody.ILMerge` to merge the dependencies into your DotNet dll - that way you only have two files, the Native DLL and the DotNet DLL

**Comment by ThaDaVos** [comment]
Did you include your `runtime.config.json` in the directory of MetaTrader? If yes, can you share the contents?

**Comment by DennisGundersen** [comment]
> About the dependencies - this, as far as I know, has always been this way - as the default DotNet install does not contain Nuget Package dll's and stuff.

Yes, that sounds right, I was just musing that things might have changed with .NET Core and hence it could open up even more venues for possible errors.

I have six suggested error sources right now, but of course I'm just pulling them out of my butt for a hail Mary...
Potential bugs:
1. MT4 doesn't find the C# method (naming?)
2. Marshalling (not blittable even for int only?)
3. MT4 can't read the installed .NET libraries? 
4. Needs "Self-contained" instead of "Framework-dependent"?
5. Can't read files in installation folder either? 
6. Some other file placement issues?

Re
D

[MT4Bridge.runtimeconfig.json.txt](https://github.com/AaronRobinsonMSFT/DNNE/files/11330819/MT4Bridge.runtimeconfig.json.txt)

**Comment by ThaDaVos** [comment]
Gonna try to answer your list:

1. Can you share code of both sides? (Maybe put the project on GitHub so we can help better)
2. What are you trying to Marshall? DNNE is setup in a way that that's not necassary
3. When deploying the generated DLL based on my template - you need to have DotNet Runtime installed for the version you used - that's what `Framework Dependent` means - I haven't tried to create a single file library yet - may try eventually when I get a good example working
4. Why needs `Self-Contained`?
5. Depends on how they are placed and which file you copied - can you show a list of the files you copied?
6. Maybe

PS: Let's not start about Clarion - it used custom formats for almost everything 🤣

**Comment by DennisGundersen** [comment]
> Gonna try to answer your list:
> 
> 1. Can you share code of both sides? (Maybe put the project on GitHub so we can help better)
> 2. What are you trying to Marshall? DNNE is setup in a way that that's not necassary
> 3. When deploying the generated DLL based on my template - you need to have DotNet Runtime installed for the version you used - that's what `Framework Dependent` means - I haven't tried to create a single file library yet - may try eventually when I get a good example working
> 4. Why needs `Self-Contained`?
> 5. Depends on how they are placed and which file you copied - can you show a list of the files you copied?
> 6. Maybe
> 

Sorry for the slow reply, I had to learn Git. I have now created a repository and tried to fill it out with the info needed about both programming in MT4 and including DNNE (as far as I've been able to gather it so far). Hadn't really thought about publishing anything in my own name, but with some info about Git, I see how this will enable easier collaboration without messing up too much for Aaron. The repository is at https://github.com/DennisGundersen/MT4_To_CSharp_Bridge.

Re
Dennis

**Comment by AaronRobinsonMSFT** [comment]
@DennisGundersen Thanks for the repo and example. Is there any possibility we can get free access to MT4 so we can run this locally?

Were you able to try using any of the approaches I mentioned at https://github.com/AaronRobinsonMSFT/DNNE/issues/165#issuecomment-1499670382?

I assume you've proved out that the 32-bit DNNE scenarios work on your machine outside of the MT4 environment.

**Comment by DennisGundersen** [comment]
> @DennisGundersen Thanks for the repo and example. Is there any possibility we can get free access to MT4 so we can run this locally?

There are massive amounts of Forex brokers that offer free demo accounts with MT4 (downloadable) for training and development, such as https://roboforex.com/. (MT4 is the locally installed client app, and the brokers run the servers which receive the market data and place orders on your behalf).
 
> Were you able to try using any of the approaches I mentioned at [#165 (comment)](https://github.com/AaronRobinsonMSFT/DNNE/issues/165#issuecomment-1499670382)?
> 
> I assume you've proved out that the 32-bit DNNE scenarios work on your machine outside of the MT4 environment.

I'm still looking to hire an actual developer over at UpWork.com (https://www.upwork.com/jobs/~0161db045c3c0e71a3) to help me with this part as [NrSlionHeart](https://github.com/AaronRobinsonMSFT/DNNE/issues?q=is%3Aissue+author%3ANrSlionHeart) hasn't gotten back to me in weeks, but since [ThaDaVos](https://github.com/ThaDaVos) offered to help if I set up a repository, I thought I might as well get a jump start on it. I'm really not expecting you to spend your time on my obscure situation though. I'm sure I'll find someone who can implement your suggestions and then ask cogent questions afterwards if needed.

Re
Dennis

**Comment by ThaDaVos** [comment]
@DennisGundersen  - I see in your repo that `SelfContained` is set to `true` - as far as I know this should be `false`
My current `.csproj` looks as follows:
```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup Label="Project">
    <TargetFramework>net7.0</TargetFramework>
    <Nullable>enable</Nullable>
    <EnableDynamicLoading>true</EnableDynamicLoading>
    <OutputType>library</OutputType>
    <AutoGenerateBindingRedirects>true</AutoGenerateBindingRedirects>
    <InvariantGlobalization>true</InvariantGlobalization>
    <CopyLocalLockFileAssemblies>true</CopyLocalLockFileAssemblies>
  </PropertyGroup>
  <Target Name="Change DNNE Settings" BeforeTargets="DnneGenerateNativeExports">
    <PropertyGroup Label="DNNE">
      <DnneRuntimeIdentifier>win-x86</DnneRuntimeIdentifier>
      <DnneSelfContained_Experimental>true</DnneSelfContained_Experimental>
      <DnneAddGeneratedBinaryToProject>true</DnneAddGeneratedBinaryToProject>
      <DnneAdditionalGenerators>Clarion*|XML*|XSD*</DnneAdditionalGenerators>
    </PropertyGroup>
    <Message Text="Configuring DNNE using a task" />
  </Target>
  <PropertyGroup Label="Publish">
    <GenerateRuntimeConfigurationFiles>true</GenerateRuntimeConfigurationFiles>
    <RollForward>LatestMinor</RollForward>
    <PublishSingleFile>false</PublishSingleFile>
    <SelfContained>false</SelfContained>
    <PublishTrimmed>false</PublishTrimmed>
    <DebugType>embedded</DebugType>
    <IncludeNativeLibrariesForSelfExtract>false</IncludeNativeLibrariesForSelfExtract>
    <ErrorOnDuplicatePublishOutputFiles>false</ErrorOnDuplicatePublishOutputFiles>
  </PropertyGroup>
   <!-- PackageReferences etc -->
</Project>
```

You can leave the `Target` out - and just use the `PropertyGroup` inside as you do now - I was using a `Target` as I also modified my local DNNE for some extra functionality's like auto-generation of wrapping code in Clarion - see: https://github.com/AaronRobinsonMSFT/DNNE/issues/162

**Comment by DennisGundersen** [comment]
> @DennisGundersen - I see in your repo that `SelfContained` is set to `true` - as far as I know this should be `false` My current `.csproj` looks as follows:

I've tried both, but that's not where the problem is apparently. Now I'm trying to setup a C++ project in VS and load either the C# project or the dll, but I can't find any good info on that either. All examples are the other way. Still haven't found a C++ developer to take on the project. And the weeks go by...

Re
Dennis

**Comment by AaronRobinsonMSFT** [comment]
@DennisGundersen You should be able to reasonably generate or load the following into Visual Studio - https://github.com/AaronRobinsonMSFT/DNNE/tree/master/test/ImportingProcess. From there you can likely edit what is needed for your scenario. Let me know if you need some help editing the existing `main()`.

**Comment by DennisGundersen** [comment]
Yeah, so... The thing is... It turns out that if you include both your own code (ProjectName.dll), DNNE (ProjectNameNE.dll) AND the .NET configuration (ProjectName.runtimeconfig.json) in the /MQL/Library folder, it works a lot better. Sorry about that.
Complete prototype for MT4 to C# is now available at https://github.com/DennisGundersen/MT4_To_CSharp_Bridge.

Re
Dennis
